File: app/src/main/java/com/aso/senote/data/models/Note.java
/*
 * Copyright (c) ASO 2020.
 */

package com.aso.senote.data.models;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.Index;
import androidx.room.PrimaryKey;

import java.util.Calendar;

@Entity(tableName = "note_table",
        indices = {@Index(value = "noteId"), @Index(value = "notebookId")},
        foreignKeys = @ForeignKey(entity = Notebook.class,
                parentColumns = "notebookId",
                childColumns = "notebookId",
                onDelete = ForeignKey.CASCADE)
)
public class Note {


    @ColumnInfo(name = "date_created")
    private Calendar dateCreated;
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "noteId")
    private long id;
    private long notebookId;
    private String noteTitle;
    private String noteContent;
    private Calendar lastModified;

    public Note(long notebookId, String noteTitle, String noteContent) {
        this.dateCreated = Calendar.getInstance();
        this.notebookId = notebookId;
        this.noteTitle = noteTitle;
        this.noteContent = noteContent;
        this.lastModified = Calendar.getInstance();
    }

    public long getNotebookId() {
        return notebookId;
    }

    public void setNotebookId(long notebookId) {
        this.notebookId = notebookId;
    }

    public String getNoteTitle() {
        return noteTitle;
    }

    public void setNoteTitle(String noteTitle) {
        this.noteTitle = noteTitle;
    }

    public String getNoteContent() {
        return noteContent;
    }

    public void setNoteContent(String noteContent) {
        this.noteContent = noteContent;
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public Calendar getDateCreated() {
        return dateCreated;
    }

    public void setDateCreated(Calendar dateCreated) {
        this.dateCreated = dateCreated;
    }

    public Calendar getLastModified() {
        return lastModified;
    }

    public void setLastModified(Calendar lastModified) {
        this.lastModified = lastModified;
    }
}

File: app/src/main/java/com/aso/senote/data/models/Note.kt
/*
 * Copyright (c) ASO 2020.
 */
package com.aso.senote.data.models

import androidx.room.*
import java.util.*

@Entity(tableName = "note_table",
        indices = [Index(value = ["noteId"]),
            Index(value = ["notebookId"])],
        foreignKeys = [
            ForeignKey(entity = Notebook::class,
                    parentColumns = ["notebookId"],
                    childColumns = ["notebookId"],
                    onDelete = ForeignKey.CASCADE)
        ])
class Note {
    @ColumnInfo(name = "date_created")
    var dateCreated: Calendar

    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "noteId")
    var id: Long = 0
    var notebookId: Long = 0
    var noteTitle: String? = null
    var noteContent: String? = null
    var isStarred = false
    var lastModified: Calendar

    constructor(notebookId: Long, noteTitle: String?, noteContent: String?) {
        dateCreated = Calendar.getInstance()
        this.notebookId = notebookId
        this.noteTitle = noteTitle
        this.noteContent = noteContent
        lastModified = Calendar.getInstance()
        isStarred = false
    }

    @Ignore
    constructor() {
        dateCreated = Calendar.getInstance()
        lastModified = Calendar.getInstance()
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o !is Note) return false
        val note = o
        return id == note.id && notebookId == note.notebookId && isStarred == note.isStarred && dateCreated == note.dateCreated &&
                noteTitle == note.noteTitle &&
                noteContent == note.noteContent && lastModified == note.lastModified
    }

    override fun hashCode(): Int {
        return Objects.hash(dateCreated, id, notebookId, noteTitle, noteContent, isStarred, lastModified)
    }
}